fitbenchmarking/parsing/sscanss_parser.py

```diff
@@ -0,0 +1,131 @@
+"""
+import math
+import os
+import sys
+import typing
+from importlib import import_module
+
+import numpy as np
+
+from fitbenchmarking.parsing.fitbenchmark_parser import FitbenchmarkParser
+from fitbenchmarking.parsing.fitting_problem import FittingProblem
+
+
+class SScanSSParser(FitbenchmarkParser):
+    """
+    Parser for a SScanSS problem definition file.
+
+    Function calculates the error from a current pose to the target pose.
+    """
+    PARAM_IGNORE_LIST = ['robot', 'module', 'targets']
+
+    def parse(self) -> 'list[FittingProblem]':
+        template = super().parse()
+
+        out = []
+
+        pf = self._parsed_func[0]
+        path = os.path.join(os.path.dirname(self._filename), pf['module'])
+        sys.path.append(os.path.dirname(path))
+        module = import_module(os.path.basename(path))
+        robot = getattr(module, pf['robot'])
+        targets = getattr(module, pf['targets'])
+
+        for i, t in enumerate(targets):
+            fp = FittingProblem(self.options)
+            fp.multifit = template.multifit
+            fp.name = f'{template.name} - target {i+1}'
+            fp.description = (f'{template.description.rstrip().rstrip(".")}. '
+                              f'This problem is associated with target {i+1}.')
+            fp.function = self.inverse_kinematics_error(robot, t)
+            fp.format = template.format
+            fp.plot_scale = template.plot_scale
+            fp.multivariate = template.multivariate
+            fp.equation = template.equation
+            fp.starting_values = template.starting_values
+            fp.value_ranges = template.value_ranges
+            fp.data_x = template.data_x
+            fp.data_y = template.data_y
+            fp.data_e = template.data_e
+            fp.start_x = template.start_x
+            fp.end_x = template.end_x
+            out.append(fp)
+
+        return out
+
+    def _create_function(self) -> typing.Callable:
+        """
+        Required for base class.
+        Since this is only called to build the template, we return a dummy
+        function.
+
+        Expected function format:
+        function='module=robots,robot=model1,target=target1,p1=...,p2=...,etc'
+
+        :return: A callable function
+        :rtype: callable
+        """
+        # pylint: disable=unused-argument
+        def dummy(x, *params):
+            pass
+
+        return dummy
+
+    @staticmethod
+    def inverse_kinematics_error(robot, target):
+        """
+        Create a function for calculation the error in a forward kinematics
+        transform.
+        """
+        def error_func(x, *params):
+            # pylint: disable=unused-argument
+            cur_pose = robot.fkine(params)
+            pos_diff = cur_pose[:3, 3] - target[:3, 3]
+            angle_diff = SScanSSParser.emap(
+                target[:3, :3] @ cur_pose[:3, :3].transpose())
+            return np.array(list(pos_diff) + list(angle_diff))
```

Comment:
```
    current_pose = ri.fkine(q)
    error = np.zeros(6)
    error[:3] = target_pose[:3, 3] - current_pose[:3, 3]
    
    diff = target_pose[:3, :3] @ current_pose[:3, :3].transpose()
    error[3:6] = matrix_to_emap(diff)

    err = np.dot(error, error)
    return err
```

This is the snippet from the original [code](https://bitbucket.org/SteveNneji/demo-for-fitbenchmarking/src/master/main.py). Is the dot product computed outside the function ? 

I ran it once just to check the output (Like Stephen did earlier).

<img width="439" alt="image" src="https://github.com/fitbenchmarking/fitbenchmarking/assets/47083562/ef0ef212-88de-4a39-ad3c-058811f8ea96">

Some errors are too small and some are large. Some problems are exactly solvable (Two link - Target 1 and 2). But the error computed is not the same for exactly solvable problems. Would this be expected ?

Reply:
I think all of the problems are exactly solvable, in the global sense. It could be that there are local minima that some solvers get stuck in.

The dot product is done as part of the cost function:
$$c = \Vert f(x; p) - y \Vert^2$$

For this problem, the error $f(x; p) - y$ is defined by the `error_func`. Since it's in a norm, the outcome is the same.

For some solver this makes a big difference as they can get a gradient for each of the error terms.

I'll rerun it with the solvers listed. It could be that bounds would help as you mention and they should probably all be bounded (or at least the linear actuators should be!). Running with `scipy_ls` gives good results which is what I was testing with. From an initial observation, `scipy` takes much longer than `scipy_ls`.

Reply:
Looking at the graphs from the `scipy` runs, the accuracy seems to be way off for some of them. I'll have a look for why.

Reply:
@RabiyaF can you try again with an unweighted cost function. Since the weighting is being set to $\max(\sqrt{y},10^-8)$ and $y=0$, it's just scaling up the chisq for no reason. I think I'll have to explicitly set $e=0$.

Reply:
I ran it with `scipy_ls` and `nlls`. In this case, you are right, the errors are reasonable. Maybe `scipy: Nelder-Mead` is just not suitable for this problem. SScanSS uses the `nlopt` package. I would just like to verify that nlopt minimisers are giving good results. I will try to do this soon if I can. But I was having some trouble with installing `nlopt` on a Mac. But I might look into it later this week.  

<img width="747" alt="image" src="https://github.com/fitbenchmarking/fitbenchmarking/assets/47083562/0f9d5aae-b6f1-4905-b747-bbba4c179dee">

Reply:
There're quite a few minimizers for nlopt. Here's what I get on the first few:
![image](https://github.com/fitbenchmarking/fitbenchmarking/assets/48282025/ae8e6f8a-eaf6-493d-ad48-769a9f0893f9)